src/protocols/output_management.rs

```diff
@@ -0,0 +1,785 @@
+    D: OutputManagementHandler,
+    D: 'static,
+{
+    fn request(
+        state: &mut D,
+        client: &Client,
+        conf_head: &ZwlrOutputConfigurationHeadV1,
+        request: zwlr_output_configuration_head_v1::Request,
+        conf: &ZwlrOutputConfigurationV1,
+        _display: &DisplayHandle,
+        _data_init: &mut DataInit<'_, D>,
+    ) {
+        let g_state = state.output_management_state();
+        let client_data = match g_state.clients.get_mut(&client.id()) {
+            Some(data) => data,
+            _ => {
+                error!("ConfigurationHead: missing client data");
+                return;
+            }
+        };
+        let conf_data = match client_data.confs.get_mut(conf) {
+            Some(data) => data,
+            None => {
+                error!("ConfigurationHead: missing conf data");
+                return;
+            }
+        };
+        let output_name = match conf_data.heads.get(conf_head) {
+            Some(name) => name,
+            _ => {
+                error!("ConfigurationHead: missing head data");
+                return;
+            }
+        };
+        let new_out = match conf_data.new_out.get_mut(output_name) {
+            Some(out) => out,
+            None => {
+                error!("ConfigurationHead: missing matching head conf data");
+                return;
+            }
+        };
+        if conf_data.serial != g_state.serial {
+            return;
+        }
+        match request {
+            zwlr_output_configuration_head_v1::Request::SetMode { mode } => {
+                let index = match client_data
+                    .heads
+                    .get(output_name)
+                    .map(|(_, mods)| mods.iter().position(|m| m.id() == mode.id()))
+                {
+                    Some(Some(index)) => index,
+                    _ => {
+                        error!("SetMode: failed to find requested mode");
+                        return;
+                    }
+                };
+                if new_out.modes.len() <= index {
+                    error!("SetMode: requested mode is out of range");
+                    return;
+                }
+                new_out.current_mode = Some(index);
+            }
+            zwlr_output_configuration_head_v1::Request::SetCustomMode {
+                width,
+                height,
+                refresh,
+            } => {
+                let (width, height, refresh): (u16, u16, u32) =
+                    match (width.try_into(), height.try_into(), refresh.try_into()) {
+                        (Ok(width), Ok(height), Ok(refresh)) => (width, height, refresh),
+                        _ => {
+                            error!("SetCustomMode: invalid input data");
+                            return;
+                        }
+                    };
+                if let Some(index) = new_out.modes.iter().position(|m| {
```

Comment:
I guess what this request is intended to do is more like https://github.com/YaLTeR/niri/issues/333 rather than picking one of the existing modes? I suppose this logic is fine until that's implemented.